pkg/registry/populator.go

```diff
@@ -0,0 +1,256 @@
+		if err != nil {
+			return fmt.Errorf("unable to read file %s: %s", f.Name(), err)
+		}
+		decoder := yaml.NewYAMLOrJSONDecoder(fileReader, 30)
+		err = decoder.Decode(&annotationsFile)
+		if err != nil || *annotationsFile == (AnnotationsFile{}) {
+			continue
+		} else {
+			log.Info("found annotations file searching for csv")
+		}
+	}
+
+	if *annotationsFile == (AnnotationsFile{}) {
+		return fmt.Errorf("Could not find annotations.yaml file")
+	}
+
+	err = i.loadManifests(manifests, annotationsFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *DirectoryPopulator) loadManifests(manifests string, annotationsFile *AnnotationsFile) error {
+	log := logrus.WithFields(logrus.Fields{"dir": i.from, "file": manifests, "load": "bundle"})
+
+	csv, err := i.findCSV(manifests)
+	if err != nil {
+		return err
+	}
+
+	if csv.Object == nil {
+		return fmt.Errorf("csv is empty: %s", err)
+	}
+
+	log.Info("found csv, loading bundle")
+
+	// TODO: Check channels against what's in the database vs in the bundle csv
+
+	bundle, err := loadBundle(csv.GetName(), manifests)
+	if err != nil {
+		return fmt.Errorf("error loading objs in directory: %s", err)
+	}
+
+	if bundle == nil || bundle.Size() == 0 {
+		return fmt.Errorf("no bundle objects found")
+	}
+
+	// set the bundleimage on the bundle
+	bundle.BundleImage = i.ref
+
+	if err := bundle.AllProvidedAPIsInBundle(); err != nil {
+		return fmt.Errorf("error checking provided apis in bundle %s: %s", bundle.Name, err)
+	}
+
+	bcsv, err := bundle.ClusterServiceVersion()
+	if err != nil {
+		return fmt.Errorf("error getting csv from bundle %s: %s", bundle.Name, err)
+	}
+
+	packageManifest, err := translateAnnotationsIntoPackage(annotationsFile, bcsv)
+	if err != nil {
+		return fmt.Errorf("Could not translate annotations file into packageManifest %s", err)
+	}
+
+	if err := i.loadOperatorBundle(packageManifest, *bundle); err != nil {
+		return fmt.Errorf("Error adding package %s", err)
+	}
+
+	// Finally let's delete all the old bundles
+	if err = i.loader.ClearNonDefaultBundles(packageManifest.PackageName); err != nil {
+		return fmt.Errorf("Error deleting previous bundles: %s", err)
+	}
+
+	return nil
+}
+
+// loadBundle takes the directory that a CSV is in and assumes the rest of the objects in that directory
+// are part of the bundle.
+func loadBundle(csvName string, dir string) (*Bundle, error) {
+	log := logrus.WithFields(logrus.Fields{"dir": dir, "load": "bundle"})
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var errs []error
+	bundle := &Bundle{}
+	for _, f := range files {
+		log = log.WithField("file", f.Name())
+		if f.IsDir() {
+			log.Info("skipping directory")
+			continue
+		}
+
+		if strings.HasPrefix(f.Name(), ".") {
+			log.Info("skipping hidden file")
+			continue
+		}
+
+		log.Info("loading bundle file")
+		path := filepath.Join(dir, f.Name())
+		fileReader, err := os.Open(path)
```

Comment:
Same here. I don't know what we would consider a large number of files in a bundle, but it could be worthwhile to either call Close outside of a defer or wrap the loop body and defer Close to limit peak open file descriptors. I ran into a problem with FD exhaustion [here](https://github.com/operator-framework/operator-registry/commit/626af900865653e4318855776c84868b038a10ee#diff-9521fe6a08ae105c7acc7037e3369db5), granted, the number of records in one of those archives is probably significantly higher than any sane bundle directory.